timemachine/fe/interaction_group_traj.py

```diff
@@ -0,0 +1,113 @@
+from typing import Callable
+
+import numpy as np
+from jax import jit
+from jax import numpy as jnp
+from jax import vmap
+from numpy.typing import NDArray as Array
+
+from timemachine.potentials import nonbonded
+from timemachine.potentials.jax_utils import distance2
+
+Position = Array
+Param = Array
+Box = Array
+Energy = float
+PairFxn = Callable[[Position, Position, Param, Param, Box], Energy]
+
+
+# example pair function
+def nb_pair_fxn(x_a, x_b, param_a, param_b, box):
+    beta = 2.0
+    cutoff = 1.2
+
+    # alchemical distance
+    r2 = distance2(x_a, x_b, box)
+    w_offset = param_b[3] - param_a[3]
+    r = jnp.sqrt(r2 + w_offset**2)
+
+    # TM reaction field
+    q_prod = param_a[0] * param_b[0]
+    e_q = nonbonded.switched_direct_space_pme(r, q_prod, beta, cutoff)
+
+    # Lennard-Jones
+    sig = nonbonded.combining_rule_sigma(param_a[1], param_b[1])
+    eps = nonbonded.combining_rule_epsilon(param_a[2], param_b[2])
+    e_lj = nonbonded.lennard_jones(r, sig, eps)
+
+    return jnp.where(r < cutoff, e_q + e_lj, 0.0)
+
+
+# brute-force pre-computed neighborlist
+@jit
+def env_mask_within_cutoff(x_env, x_lig, box, cutoff):
+    """result[i] = True if any distance(x_env[i], y) < cutoff for y in x_lig"""
+
+    def d2_others(x_i, x_others):
+        d2_ij = vmap(distance2, (None, 0, None))(x_i, x_others, box)
+        return jnp.where(d2_ij <= cutoff**2, d2_ij, jnp.inf)
+
+    def within_cutoff(point):
+        return jnp.any(d2_others(point, x_lig) < cutoff**2)
+
+    return vmap(within_cutoff)(x_env)
+
+
+class InteractionGroupTraj:
+    def __init__(self, pair_fxn: PairFxn, xs: Array, box_diags: Array, ligand_idxs: Array, env_idxs: Array, cutoff=1.2):
```

Comment:
I think it'd also be useful to have a serialize (to dict) and deserialize classmethod. This will let us store the state in an npz (instead of needing to pickle the object).